_posts/2021-03-02-faster-pipelines-with-knapsack-pro-by-parallelizing-cypress-tests-on-gitlab-ci.md

```diff
@@ -0,0 +1,279 @@
+      <figure>
+        <img src="/images/blog/posts/faster-pipelines-with-knapsack-pro-by-parallelizing-cypress-tests-on-gitlab-ci/08-pipeline-runtime-after-knapsack.png" alt="Pipeline summary showing: 33 jobs for master in 15 minutes and 21 seconds" />
+        <figcaption>Pipeline summary</figcaption>
+      </figure>
+    </a>
+  </article>
+</section>
+
+It is still not perfectly balanced but with this granularity, we can utilize four runners at the same time and there isn’t too much idle time, so we’re not wasting CPU cycles.
+
+<p>
+  <a target="_blank" rel="noopener" href="/images/blog/posts/faster-pipelines-with-knapsack-pro-by-parallelizing-cypress-tests-on-gitlab-ci/09-after_cypress_new_master_fullpage.png">
+    <figure>
+      <img src="/images/blog/posts/faster-pipelines-with-knapsack-pro-by-parallelizing-cypress-tests-on-gitlab-ci/09-after_cypress_new_master_fullpage.png" alt="Knapsack Pro dashboard showing the slowest test files, slowest being 21.803 seconds while the second slowest being 21.001 seconds (some other tests are only 1-6 seconds short)" />
+      <figcaption>Knapsack Pro dashboard showing the slowest test files, slowest being 21.803 seconds while the second slowest being 21.001 seconds (some other tests are only 1-6 seconds short)</figcaption>
+    </figure>
+  </a>
+</p>
+
+
+# What else can we do?
+
+15 minutes still feels long, pipeline runtimes around 10 minutes are a good goal to reach but we have more optimization to do around build times and Docker compilation times, which are out of scope for this article. If you’re interested in reading more about how to do that, I can recommend the article [Docker for JavaScript Devs: How to Containerize Node.js Apps Efficiently](https://www.robincussol.com/docker-for-js-devs-how-to-containerise-nodejs-apps-efficiently/) for more details.
+
+
+# Results
+
+The switch to Knapsack Pro took less than a day. Most of the work afterward was to split up tasks and optimize other parts of GitLab CI. The whole process took two weeks for a single developer.
+
+By splitting up the test suites into smaller, equal sized chunks, running tests parallelized on four CI runners (for each of our target builds) and some other general pipeline fixes (better caching between jobs/steps, only running parts of the pipeline for actual changes), we’ve managed to decrease our pipeline runtime of ~20 minutes to ~15 minutes.
+
+When we add more test suites, Knapsack Pro will take care of sorting the tests to the proper CI runners to ensure that pipeline runtimes don’t increase if necessary.
+
+We sped up our pipeline by 33%, or 5 minutes on average, and seeing that each of our three developers starts about 6-10 pipelines a day (depending on the workload and type of work), we’ve essentially saved between 90-150 minutes of development time a day, as well as increased morale because **_nobody likes to wait._**
```

Comment:
Hate to correct this one :smile: , but isn't going from 20 to 15 technically a 25% reduction?

Reply:
🤡

Reply:
so we went from 133% to 100%, which was confused me.

we actually went from 100% time to 75% =)

Reply:
BTW, now if your pipeline timing was to increase by 33%, you'd be back to 20 minutes. :smile: